pkg/qpoption/qpoption.go

```diff
@@ -0,0 +1,189 @@
+var qpExtensionPrefix = "qpextension.knative.dev/"
+
+type GateQPOption struct {
+	config           map[string]string
+	activated        bool
+	defaults         *sharedmain.Defaults
+	securityPlug     pi.RoundTripPlug
+	nextRoundTripper http.RoundTripper // the next roundtripper
+}
+
+func NewGateQPOption() *GateQPOption {
+	return new(GateQPOption)
+}
+
+func (p *GateQPOption) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			pi.Log.Warnf("Recovered from panic during RoundTrip! Recover: %v\n", recovered)
+			pi.Log.Debugf("Stacktrace from panic: \n %s\n" + string(debug.Stack()))
+			err = errors.New("panic during RoundTrip")
+			resp = nil
+		}
+	}()
+
+	if req, err = p.securityPlug.ApproveRequest(req); err != nil {
+		pi.Log.Debugf("%s: returning error %v", p.securityPlug.PlugName(), err)
+		resp = nil
+		return
+	}
+
+	if resp, err = p.nextRoundTripper.RoundTrip(req); err == nil {
+		resp, err = p.securityPlug.ApproveResponse(req, resp)
+	}
+
+	return
+}
+
+func (p *GateQPOption) ProcessAnnotations(annotationsPath string, qpExtPrefix string) bool {
+	file, err := os.Open(annotationsPath)
+	if err != nil {
+		p.defaults.Logger.Debugf("File %s is can not be opened - is PodInfo mounted? os.Open Error: %s", annotationsPath, err.Error())
+		return false
+	}
+	defer file.Close()
+	p.config = make(map[string]string)
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		txt := scanner.Text()
+		txt = strings.ToLower(txt)
+		parts := strings.Split(txt, "=")
+
+		k := parts[0]
+		v := parts[1]
+		if strings.HasPrefix(k, qpExtPrefix) && len(k) > len(qpExtPrefix) {
+			v = strings.TrimSuffix(strings.TrimPrefix(v, "\""), "\"")
+			k = k[len(qpExtPrefix):]
+			keyParts := strings.Split(k, "-")
+			if len(keyParts) < 2 {
+				continue
+			}
+			extension := keyParts[0]
+			action := keyParts[1]
+			if strings.EqualFold(extension, p.securityPlug.PlugName()) {
+				switch action {
+				case "activate":
+					if strings.EqualFold(v, "enable") {
+						p.activated = true
+					}
+				case "config":
+					if len(keyParts) == 3 {
+						extensionKey := keyParts[2]
+						p.config[extensionKey] = v
+					}
+				}
+			}
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		p.defaults.Logger.Infof("File %s - scanner Error %s", annotationsPath, err.Error())
+		return false
+	}
+	return true
+}
+
+func (p *GateQPOption) Setup(defaults *sharedmain.Defaults) {
+	// Never panic the caller app from here
+	defer func() {
+		if r := recover(); r != nil {
+			pi.Log.Warnf("Recovered from panic during Setup()! Recover: %v", r)
+		}
+	}()
+
+	if pi.RoundTripPlugs == nil || len(pi.RoundTripPlugs) == 0 {
+		pi.Log.Warnf("Image was created with qpoption package but without a Plug")
+		return
+	}
+	if len(pi.RoundTripPlugs) > 1 {
+		pi.Log.Warnf("Image was created with more then one plug")
+		return
+	}
```

Comment:
The rtplugs package supports multiple plugs
I avoided it since it was a bit too much for this repo. 

Originally, the plan was for two repos - one for generic extendibility of QP with qpoption, rtplugs, testgate and pluginterfaces , and another for security-related codes

Since we start with a single repo, I am attempting to remove the support for multiple plugs to somewhat reduce the code complexity - while assuming that we will move back to rtplugs later on.  

>> I will add a comment

Reply:
Solved in pluginterfaces  - see GetPlug()